internal/blockchain/difficulty.go

```diff
@@ -212,20 +211,152 @@ func (b *BlockChain) calcNextRequiredDifficulty(prevNode *blockNode, newBlockTim
+	//
+	// Smaller numbers result in a higher difficulty, so imposing a maximum
+	// difficulty equates to limiting the minimum target value.
+	if b.minTestNetTarget != nil && nextDiff < b.minTestNetDiffBits {
+		nextDiff = b.minTestNetDiffBits
+	}
+
+	return nextDiff
+}
+
+// blake3WorkDiffAnchor returns the block to treat as the anchor block for the
+// purposes of determining how far ahead or behind the ideal schedule the
+// provided block is when calculating the blake3 target difficulty for the block
+// AFTER the passed previous block node.
+//
+// This function MUST only be called with the blake3 proof of work agenda active
+// and with the chain state lock held (for writes).
+func (b *BlockChain) blake3WorkDiffAnchor(prevNode *blockNode) *blockNode {
+	// Use the previously cached anchor when it exists and is actually an
+	// ancestor of the passed node.
+	//
+	// Note that it will always be set to the genesis block on networks where
+	// the agenda is always active.
+	anchor := b.cachedBlake3WorkDiffAnchor.Load()
+	if anchor != nil && anchor.IsAncestorOf(prevNode) {
+		return anchor
+	}
+
+	// Find the block just prior to the activation of the blake3 proof of work
+	// agenda from the perspective of the block AFTER the passed block.
+	//
+	// The state of an agenda can only change at a rule change activation
+	// boundary, so determine the final block of the rule change activation
+	// interval just prior to the interval the block AFTER the provided one
+	// falls in and then loop backwards one interval at a time until the agenda
+	// is no longer active.
+	rcai := int64(b.chainParams.RuleChangeActivationInterval)
+	svh := b.chainParams.StakeValidationHeight
+	finalNodeHeight := calcWantHeight(svh, rcai, prevNode.height+1)
+	candidate := prevNode.Ancestor(finalNodeHeight)
+	for candidate != nil && candidate.parent != nil {
+		// Since isBlake3PowAgendaActive returns the state for the block AFTER
+		// the provided one and the goal here is to determine the state of the
+		// agenda for the candidate anchor (which is the final block of the rule
+		// change activation interval under test), use the parent to get the
+		// state of the candidate itself.
+		isActive, err := b.isBlake3PowAgendaActive(candidate.parent)
+		if err != nil {
+			panicf("known good agenda state lookup failed for block node "+
+				"hash %v (height %v) -- %v", candidate.parent.hash,
+				candidate.parent.height, err)
+		}
+		if !isActive {
+			anchor = candidate
+			break
+		}
+
+		// Skip back to the ancestor that is the final block of the previous
+		// rule change interval.
+		candidate = candidate.RelativeAncestor(rcai)
+	}
+
+	// Update the cached anchor to the discovered one since it is highly likely
+	// the next call will involve a descendant of this anchor as opposed to some
+	// other anchor on an entirely unrelated side chain.
+	if anchor != nil {
+		b.cachedBlake3WorkDiffAnchor.Store(anchor)
```

Comment:
This is the only place (other than at startup when the agenda is forced) where this is set. 

There could be some flickering of this value, right after the activation block, if multiple sidechains end up being mined, but this should be resolved fairly quickly (and in any case this func will return very fast, because the first iteration of the loop will already succeed).

This is also the reason why the access to `cachedBlake3WorkDiffAnchor` at the top of this function is gated by checking whether `anchor.isAncestor(prevNode)`.